### ISSUE TITLE
Add compile/runtime classpath for all sourceSets to Eclipse by default

### ISSUE THREAD

**Issue by staffanf** [eclipse/buildship#476]
A build with 2 subprojects, app and sub.
Both projects have added a sourceSet "sim" with a matching configuration 'sim'.
The following dependency is not resolved correctly in Eclipse.

```
dependencies {
  simCompile project(path: ':sub', configuration: 'sim')
}
```

I get compile errors in Eclipse with Buildship that are not present in CLI/IntelliJ builds.

### Expected Behavior
Eclipse with Buildship should resolve the correct dependencies for the simCompile configuration in app.

### Current Behavior
Eclipse with Buildship does not properly resolve the sub-sim.jar from the sub project, making Eclipse fail while compiling.

### Context
Our CI builds work but the Eclipse build fails due to incorrectly resolved dependencies.
We build our simulator parts in each module and then combine these into a main application and main simulator application. This builds works in the command line and IntelliJ builds, but not in Eclipse (Mars.2/Neon.3) with Buildship (2.0.2).

### Steps to Reproduce (for bugs)
- Clone  https://github.com/staffanf/buildship-sourceset-tests
- Import into Buildship

### Your Environment
Linux, Eclipse (Mars.2/Neon.3) with Buildship (2.0.2).

**Comment by oehme** [comment]
You need to tell Eclipse about the new configuration:

```
eclipse.classpath.plusConfigurations << configurations.simCompile
```

**Comment by staffanf** [comment]
@oehme  Shouldn't Buildship do this for me? It's in the compile classpath for ':app'.

**Comment by oehme** [comment]
That's a fair point, we're adding all sourceSets by default, so we might want to add all their compile/runtime classpaths by default as well.

**Comment by staffanf** [comment]
@oehme Thanks for reconsidering.
I would argue that if a sourceSet is added automatically, then it should resolve the correct classpath(s), anything else is unexpected from the user's point of view.

**Comment by Chetic** [comment]
+1 @donat @oehme 
Any chance we will see this in next release?

**Comment by oehme** [comment]
It's not currently planned and not very high priority, since there is an easy workaround (adding the configurations explicitly). If someone is interested in providing a PR, I'll be happy to guide them through the process.

**Comment by rahulkhimasia** [comment]
I too ran into this. I add that this is happening for external (maven) dependencies of the added sourceSet too. I was able to successfully apply the suggested work around.

**Comment by markus-wa** [comment]
Has there been any progress on this?

**Comment by thc202** [comment]
@oehme I'm interested in providing a pull request. I took a look at the code and ended up with the following fix:
```diff
--- a/subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/EclipsePlugin.java
+++ b/subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/EclipsePlugin.java
@@ -26,6 +26,7 @@
 import org.gradle.api.Project;
 import org.gradle.api.Task;
 import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.ConfigurationContainer;
 import org.gradle.api.artifacts.Dependency;
 import org.gradle.api.internal.ConventionMapping;
 import org.gradle.api.internal.IConventionAware;
@@ -37,6 +38,7 @@
 import org.gradle.api.plugins.JavaPluginConvention;
 import org.gradle.api.plugins.WarPlugin;
 import org.gradle.api.plugins.scala.ScalaBasePlugin;
+import org.gradle.api.tasks.SourceSet;
 import org.gradle.api.tasks.SourceSetContainer;
 import org.gradle.api.tasks.TaskProvider;
 import org.gradle.internal.xml.XmlTransformer;
@@ -60,6 +62,7 @@
 
 import javax.inject.Inject;
 import java.io.File;
+import java.util.Collection;
 import java.util.List;
 import java.util.Set;
 import java.util.concurrent.Callable;
@@ -237,7 +240,19 @@ private static void configureJavaClasspath(final Project project, final TaskProv
         project.getPlugins().withType(JavaPlugin.class, new Action<JavaPlugin>() {
             @Override
             public void execute(JavaPlugin javaPlugin) {
-                model.getClasspath().setPlusConfigurations(Lists.newArrayList(project.getConfigurations().getByName("compileClasspath"), project.getConfigurations().getByName("runtimeClasspath"), project.getConfigurations().getByName("testCompileClasspath"), project.getConfigurations().getByName("testRuntimeClasspath")));
+                ((IConventionAware) model.getClasspath()).getConventionMapping().map("plusConfigurations", new Callable<Collection<Configuration>>() {
+                    @Override
+                    public Collection<Configuration> call() {
+                        SourceSetContainer sourceSets = project.getConvention().getPlugin(JavaPluginConvention.class).getSourceSets();
+                        List<Configuration> sourceSetsConfigurations = Lists.newArrayListWithCapacity(sourceSets.size() * 2);
+                        ConfigurationContainer configurations = project.getConfigurations();
+                        for (SourceSet sourceSet : sourceSets) {
+                            sourceSetsConfigurations.add(configurations.getByName(sourceSet.getCompileClasspathConfigurationName()));
+                            sourceSetsConfigurations.add(configurations.getByName(sourceSet.getRuntimeClasspathConfigurationName()));
+                        }
+                        return sourceSetsConfigurations;
+                    }
+                }).cache();
                 ((IConventionAware) model.getClasspath()).getConventionMapping().map("classFolders", new Callable<List<File>>() {
                     @Override
                     public List<File> call() {
```
Is it acceptable?

Added the following test to verify it works:
```diff
--- a/subprojects/ide/src/integTest/groovy/org/gradle/plugins/ide/eclipse/EclipseSourceSetIntegrationSpec.groovy
+++ b/subprojects/ide/src/integTest/groovy/org/gradle/plugins/ide/eclipse/EclipseSourceSetIntegrationSpec.groovy
@@ -51,6 +51,32 @@ class EclipseSourceSetIntegrationSpec extends AbstractEclipseIntegrationSpec {
         classpath.lib('junit-4.13.jar').assertHasAttribute('gradle_used_by_scope', 'test')
     }
 
+    @ToBeFixedForInstantExecution
+    def "Custom source set defined on dependencies"() {
+        setup:
+        buildFile << """
+            apply plugin: 'java'
+            apply plugin: 'eclipse'
+
+            ${jcenterRepository()}
+
+            sourceSets {
+                integTest
+            }
+
+            dependencies {
+                integTestImplementation 'com.google.guava:guava:18.0'
+            }
+        """
+
+        when:
+        run 'eclipse'
+
+        then:
+        EclipseClasspathFixture classpath = classpath('.')
+        classpath.lib('guava-18.0.jar').assertHasAttribute('gradle_used_by_scope', 'integTest')
+    }
+
     @ToBeFixedForInstantExecution
     def "Source sets defined on source folders"() {
         setup:
```

The `:ide:check` task builds successfully.

**Comment by thc202** [comment]
@donat ping ^

**Comment by donat** [comment]
@thc202 I think there's no reason why we should not do it. Would you mind creating a PR?

**Comment by donat** [comment]
Fixed in Gradle 6.8.
